### ISSUE TITLE
Add + and - operators for math primitives to Vector2/3/4(i)

### ISSUE THREAD

**Issue by AndreSacilotto** [godotengine/godot-proposals#5476]
### Describe the project you are working on

"Mesh" generation project, using shaders

### Describe the problem or limitation you are having in your project

I will use C# for the exemples, but this proposal it's not for C# only.

Not really a limitation, but I think is pretty annoying and memory inefficient to do "vertex + new Vector2(2f, 2f)" when could simply be "vertex + 2f".

In Unity this is allowed. ~~(Please don't comment - "Here is not unity")~~

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add operators overloads to + and - to the classes on the title.

I know it's not a totally correct mathematical definition of a vector, but it's pretty accurate if we think that vector is just a point in space, and that's what godot vectors are most often used for. And it would be align with what '*' and '/' are already used for, thought they are mathematical correct.

So the pros are:

1. In shaders you can do it, adding this would result in a more consistente workflow between the two.
2. You would create one/two/tree less float/int variable(s), so less memory wasted.
3. Make easier to convert code from other engines that already have it.

And the cons:

1. Maybe some math purists/lovers would be angry/upset?

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Add overload operators c# and c++ have this feature.
I will try do a PR for the c# side later

### If this enhancement will not be used often, can it be worked around with a few lines of script?

For me will be used very very often. But there is workarounds you can define a function to do that or in C# you can make an extension method. 
But is a workround will not result in a cleaner code.

### Is there a reason why this should be core and not an add-on in the asset library?

Need to modify internal engine classes, so not happening.

**Comment by Mickeon** [comment]
> Maybe some math purists/lovers would be angry/upset?

I do not understand the cons. This looks like a straight enhancement to me.

**Comment by Mickeon** [comment]
> Maybe some math purists/lovers would be angry/upset?

It do not believe it's because of this. The reason is very similar as to why no **Vector\*** constructor accepting a single parameter is provided in Godot, unlike similar languages (e.g. `(Vector3(7.5)`).
It can be especially easy to mix-up and confuse, for readers and beginners alike. You personally, along with many, may not believe as such because you may have gotten accustomed to it, and that is pretty nice. However, Godot prioritises being clear and explicit whenever reasonable. 

For example, suppose I, _beginner_, want to nudge my character 2 pixels to the right and 2 pixels down. I write this:
```GDScript
var offset = Vector2(2, 2)
# [...]
position = position + horizontal_move.x
```
And then it doesn't compile. Why? Because I mixed up **Vector2** and **numbers** together. Oops.
If it were to work as suggested in this proposal, it would compile completely fine, then the moment I decide to nudge my character 5 pixels down instead, it would behave strangely, I'd lose a few screws over it, and then realise my miniscule yet troublesome mistake.

So, ideally, it would be best if the operation were to error out instead of misbehave.


Although, an argument in favour can be made for the Shader language supporting it.

> ### If this enhancement will not be used often, can it be worked around with a few lines of script?
You can also do `vertex + Vector2.ONE * 2`, at least in GDScript. Not sure how constants are defined in C#, but the syntax should be similar. I don't think you should worry about performance, the compiler should be smarter than that.

**Comment by Calinou** [comment]
I also don't think allowing Vector2 operators to work with float/int is a good idea, as it's too error-prone. It's better to have to type some more characters and avoid difficult-to-debug errors this way.

**Comment by AndreSacilotto** [comment]
@Calinou 
> I also don't think allowing Vector2 operators to work with float/int is a good idea, as it's too error-prone. It's better require typing some more characters and avoid difficult-to-debug errors this way.

Also dont think is a good ideia to have implicity convertions of int to float or the reverse, so would be only float for Vector2/3/4 and int for Vector2i/3i/4i.
This is the reason I make the PR, was to show  that. But I'm not sure I interpreted your comment correctly.

> difficult-to-debug errors

So this ability can be remove from shaders has they are far harder for debugging.
I googled a bit on #shader discord to see if anyone had this problem in the past and found nothing. 

@Mickeon

> About Vector constructor with one argument

I'm not gonna line I want that, but I don't see why to add it:
1. Reasonable
2. The compiler could simply inline the extension function with no problems.
3. Reasonable
4. Arguable

> Your workround

The workround you present is valid (even in c#), but I dont belive the compiler can optimize more than this:
"vertex + Vector2(2f, 2f)", because not using vec2 could change the behavior of the operation, so you would still wasting memory. And in vector operation I think speed matters.

> **About Beginners**

> Because I mixed up Vector2 and numbers together 

Well multiplication and division exist. But I get what was truly misunderstand in that code was machine logic vs human logic.

> About the Error

I think it's good not error is given, it makes you understand what the code does instead of believing what it does.

Maybe you are too worried about an error that rarely happens, because only 2 understanding are possible:
A. The X is add both Vector.X and Vector.Y;
B. The X is add to Vector.X only;  _But this one has a logical flaw, because X could be added to Vector.Y instead_

X and Y are human logic, any programmer should now that. And if they dont know is good time to understant it.

So people that know how to code would notice how B makes no sense and would assume A is right.

**Comment by AndreSacilotto** [comment]
```
position = vec2(4, 5)
motion = vec2(2, 3)
position = position + motion .x
position = position * motion .x
```
What is the difference between line 1 and 2? Only the operator.

Why would you assume that in the add operation the result would be `vec2(6, 5)` when in the result of the mult is `vec2(8, 10)` this is inconsistent. The logical result of the add operation is `vec2(6, 7)`.

@Mickeon From your logic the result of the mult should be `vec2(8, 5)` or an error, so why it's not?

@Calinou Why multiplication is OK, when add is not? Because a book or your teacher said?

You guys are the one that have premises in your heads, and cannot see it. This is why you two are math purists, you are only looking from math side, take a look from logic side and you will see what I see.

The current way only benefit the people how learn math from books, and not the one which have learned from code pratice like myself.

**Comment by Mickeon** [comment]
Remember that [Code of Conduct](https://godotengine.org/code-of-conduct) is applied here, too.

I am not particularly educated in math. I have not been educated by anyone about Vector math. From what I'm aware of, and from what I've seen across the codebase, multiplying **Vectors** by a given float is a pretty common operation. 

I don't know if I can say the same about this Proposal. Maybe in the future the need for this operation may increase and really warrant its inclusion, but in the years Godot has been public, the operation has yet to be implemented, as simple as it is, which has me wondering about its utility. It doesn't help that no concrete example on its usage has been given as of the time I'm writing this.

**Comment by AndreSacilotto** [comment]
Sorry by that im a too agressive sometimes. Have family and friends, and they say get used to it. But you guys are not used to me, sorry I mean no harm. Try to read my comments without any agressivite, I'm a very calm person in reality. But I do not pass this vibe from text, and I know that. Tried to edit some of the agressivity from comments, sorry @Mickeon.

**Comment by AndreSacilotto** [comment]
I was trying to find a good exemple and notice that `float / Vector2` also dont exist, so if this would be added this operation would also make sense to be added.

But anyway here is a exemple, one important thing ot notice this proposal do not bring any utility to the table everyting can already be done in godot current state. It is all about cleaner code.

```csharp
private Vector2 Zoom;

public void ZoomDir(float dir)
{
    const float K = 3;

    //OLD CODE
    Zoom += new Vector2(K / (K + Zoom.x), K / (K + Zoom.y)) * (dir * invertZoom);

    //NEW CODE
    Zoom += K / (Zoom + K) * dir;

    Zoom = Zoom.Clamp(MIN_ZOOM, MAX_ZOOM);
}
```
This a zoom camera2D code that uses diminishes returns. If dir > 0 it's ZoomIn if dir < 0 it's ZoomOut.

You can see how the new code would be much more compact.

**Comment by Mickeon** [comment]
How much of a boost would it even be? Without actual comparisons it's difficult to say.

**Comment by AndreSacilotto** [comment]
Hard to say even with benchmarks would be difficult to say, values types are cheap and work different depending on the OS.

Let me propose other solution, instead of overload operators what could be done is an `Add()`* method be added. So we can have that memory performance if needed.

*A Subtraction method could also be added, but would be redundant. And I dont think Add is good name others suggestions would be Sum, Addiction, AddUp.

**Comment by AndreSacilotto** [comment]
I decide to search in other engines Vector2 "source code"`or docs, to see if they would already have this feature so here is my results:

**Have Operators:** `Unity`, `Unreal`, `FlexEngine`, 

**Have Only Funcs:** 

**Don't Have Any:** `Cocos2d`, `CryEngine`, `Godot`, `Phaser`, `lumberyard`, `Monogame/XNA`, `Stride`, `Raylib`

**Don't Even Have a Vector Class:** `Constuct 3`, `GM2`

This is an funny data, so comercial engines have it while open source opens don't.

Probably because of controversial topic, so it will never have approval of a great majority so it never merged.

**Comment by zhehangd** [comment]
I fully support adding +- scalar operators

Vectors in Godot do not merely represent the ones in vector algebra, they are also general purpose multidimensional homogeneous algebraic containers for `float` and `int`. They are also commonly used to represent colors, positions (different from position vectors), and many kinds of coordinates. Rejecting basic algebraic operations like scalar addition/subtraction on the grounds that they are "undefined in vector algebra" is more trouble than it is worth.

First, no one would misinterpret the meaning of an expression like `Vector2i(1, 2) + 2`, nor is it likely to be written accidentally due to typos. Second, the scalar + and - operators have many practical use cases. Let me list a few, with comparson between with +- scalar vs without:

Normalizing a vector variable from range `[-1, 1]` to range `[0, 1]`
```gdscript
# with
v = v * 0.5 + 0.5

# without
v = v * 0.5 + Vector3(0.5, 0.5, 0.5)
```

General linear and non-linear mapping
```gdscript
# with
var eps := 1.0
var offset := 10.0
x = x / (x + eps) + offset

# without
x = x / (x + Vector3(eps, eps, eps)) + Vector3(offset, offset, offset) # WTF
```

Padding an image
```gdscript
# with
var padding_size := 1 * 2
image_size += padding_size

# without
image_size += Vector2i(padding_size, padding_size)
```

I think it is very clear that which syntax is more readable (even if we care not about the convenience for the writer).

I am a strong advocate for Godot and GDScript, yet every time I am forced to write code like the "without" examples above (including this time), I always question myself for a second why I shouldn’t switch to another engine -- this is not rhetorical, but a genuine frustration. GDScript is designed to make programming enjoyable, and this limitation directly contradicts this philosophy.

Even GDShader (GLSL) supports
```glsl
vec2 a = vec2(1., 2.);
a += 1.0;
```
This feature is only missing in GDScript.

Beyond adding scalar addition/subtraction, another possible solution is to support constructing vectors with identical elements (i.e. constant/uniform vectors), such as `Vector3(1.0)` or `Vector3.make_uniform(1.0)`. While I favour this feature too, I agree that this is much more controversial: it is prone to misinterpretation, and there is even no widely accepted name convention for it. Let's choose the scalar operators first.

A PR is made for this proposal:
[#116796](https://github.com/godotengine/godot/pull/116796)

**Comment by dalexeev** [comment]
This doesn't make sense mathematically; you can't add quantities of different dimensions (length and area, area and volume, etc.). In my opinion, the convenience isn't worth the loss of clarity. This can lead to accidental bugs.

> ```gdscript
> # with
> v = v * 0.5 + 0.5
> 
> # without
> v = v * 0.5 + Vector3(0.5, 0.5, 0.5)
> ```

You can use `coefficient * VectorX.ONE`:

```gdscript
v = v * 0.5 + 0.5 * Vector3.ONE
```

> Even GDShader (GLSL) supports

The shader language has a different purpose, a different origin (GLSL), and is fully statically typed (unlike the gradually typed GDScript).

**Comment by Ivorforce** [comment]
> This doesn't make sense mathematically; you can't add quantities of different dimensions (length and area, area and volume, etc.). 

You _can_, and it's a very common concept in tensor programming called [broadcasting](https://numpy.org/doc/stable/user/basics.broadcasting.html) (or, from a functional view, it's a monadic mapping).
This sort of operation is supported in many major math libraries like NumPy, tensorflow, PyTorch, Eigen, xtensor, and many others.
I personally use broadcasting every single day at my job — which is, fairly enough, not making Godot code. I do support the proposal, but, agreeing with dalexeev, it _is_ questionable how widely used it would be or how confusing for new developers.

**Comment by AThousandShips** [comment]
An additional potential bug situation is dynamic typing that can cause hard to identify issues, if you accidentally have a `float` in a variable and add it to a vector thinking it was a vector it now won't error any longer, breaking past expectations

**Comment by zhehangd** [comment]
> An additional potential bug situation is dynamic typing that can cause hard to identify issues, if you accidentally have a `float` in a variable and add it to a vector thinking it was a vector it now won't error any longer, breaking past expectations

Although I can't speak for others, from my experience changing a variable from vec to scalar usually means dimensionality is no longer needed. Then
```gdscript
var source := Vector3i(5, 6, 7)
var offset := Vector3i(1, 2, 3)
var result := source + offset
```
```gdscript
var source := Vector3i(5, 6, 7)
var offset := 2
var result := source + offset
```
is exactly what I need 99% of time. If any code change ought to break everything then muliplication/division bear the same sin. We can mix int and float (which I don't like very much but is ok with it), I personally don't believe mixing vec and scalar is more serious than that.

Probably because I have a bit of data processing background and get used to `numpy` so much, ops with vec and scalar is as natural as breathing to me. I really didn't expect to see so much resistance to this topic.

**Comment by AThousandShips** [comment]
> If any code change ought to break everything then muliplication/division bear the same sin.

You can't really compare multiplication and division with a scalar to addition/subtraction though they're completely different, even if it did apply which it doesn't since we don't have addition but we have multiplication

**Comment by StarryWorm** [comment]
> how confusing for new contributors

Regarding this point, we could use inlay hints, similar to the ones the LSP adds for the `:=` assignment operator. 

<img width="486" height="57" alt="Image" src="https://github.com/user-attachments/assets/4c9d9624-4383-49eb-b83b-8c5d164a9afe" />

In this case, the LSP added `Vector2`.

We could have something like:
```
var vec : *Vector2* = Vector2(1.0, 1.0)
vec += 2.0 *Vector2(2.0, 2.0)*
```
With the things in between *'s are the inlay hints. 

This way, we get the benefit of having broadcasting, but we don't sacrifice explicitness. Same UX design principle as `:=`.

**Comment by dalexeev** [comment]
> You _can_, and it's a very common concept in tensor programming called [broadcasting](https://numpy.org/doc/stable/user/basics.broadcasting.html) (or, from a functional view, it's a monadic mapping).

I'm not saying this operation is meaningless. I'm just against using the `+` and `-` operators for these purposes. It doesn't correspond to the semantics of addition/subtraction ("you can only add/subtract homogeneous values") and is an error-prone feature.

> First, no one would misinterpret the meaning of an expression like `Vector2i(1, 2) + 2`, nor is it likely to be written accidentally due to typos.

I clearly remember at least one time when the fact that this behavior is currently a compile-time error (in a typed context) helped me prevent a bug in my code as early as possible. I didn't rename anything or change the type. I simply mistakenly added a `float` variable to a vector. When you're dealing with named variables/constants rather than literals, this is less obvious. Variables named `size`, `offset`, `speed`, `velocity`, etc. can be either vectors or scalars.

There are different approaches to defining operators. Languages ​​like PHP and JavaScript attempt to define the result of an operation if it makes any sense, which sometimes leads to curious results. Other languages ​​avoid reusing the same operator for all data types, but instead define some semantics for that operator that the operation must conform to.

Just because the `Vector2 + int` overload isn't used doesn't mean it's a good idea to use it for broadcasting. By this logic, we could define `String + int` as the concatenation of a string and the string representation of an integer, and `Array + int` as appending an integer to the end of an array.

See also https://github.com/godotengine/godot/pull/66361. _At some point_, we had the operation `String + int`. However, the `int` was treated as a Unicode code point (like `char32_t`), not a string representation of a number. I.e. `65 + "1"` was equal `"A1"`, not `66` or `"651"`.

> Regarding this point, we could use inlay hints, similar to the ones the LSP adds for the `:=` assignment operator.

> This way, we get the benefit of having broadcasting, but we don't sacrifice explicitness. Same UX design principle as `:=`.

I don't think this is a valid comparison, and inlay hints are unlikely to help you avoid the error. This proposal changes the operator's behavior at runtime and removes the compile-time error in a typed context. In an untyped context, the analyzer (and therefore the editor) doesn't know what `a` and `b` are in the expression `a + b`. Not to mention that such inlay hints would be annoying inside complex expressions.

**Comment by pshaughn** [comment]
If (1,0) + 1 meant anything I'd expect it to mean (2,0), not (2,2). Implicitly broadcasting for addition just isn't a universally understood choice of semantic.
